database/database.go

```diff
@@ -59,3 +60,46 @@ func GetBlocksAfter(blockHash Hash, dataDir string) ([]Block, error) {
 
 	return blocks, nil
 }
+
+// GetBlockByHeightOrHash returns the requested block by hash or height.
+// It uses cached data in the State struct (HashCache / HeightCache)
+func GetBlockByHeightOrHash(state *State, height uint64, hash, dataDir string) (Block, Hash, error) {
+
+	blk := Block{}
+	hsh := Hash{}
+
+	key, ok := state.HeightCache[height]
+	if hash != "" {
+		key, ok = state.HashCache[hash]
+	}
+
+	if !ok {
+		if hash != "" {
+			return blk, hsh, fmt.Errorf("invalid hash: '%v'", hash)
+		}
+		return blk, hsh, fmt.Errorf("invalid height: '%v'", height)
+	}
+
+	f, err := os.OpenFile(getBlocksDbFilePath(dataDir), os.O_RDONLY, 0600)
+	if err != nil {
+		return blk, hsh, err
+	}
+	defer f.Close()
+
+	f.Seek(key, 0)
+	scanner := bufio.NewScanner(f)
+	if scanner.Scan() {
+		if err := scanner.Err(); err != nil {
+			return blk, hsh, err
+		}
+		var blockFs BlockFS
+		err = json.Unmarshal(scanner.Bytes(), &blockFs)
+		if err != nil {
+			return blk, hsh, err
+		}
+		blk = blockFs.Value
+		hsh = blockFs.Key
+	}
+
+	return blk, hsh, nil
```

Comment:
Small thing but you could make your life easier if you would return directly blockFs from this function.

Because later in the routing you have to re-create it again: 
```
writeRes(w, database.BlockFS{hash, blk})
```

So better to just return it directly.